internal/backend/runtime/omni/controllers/omni/discovery_audit.go

```diff
@@ -0,0 +1,303 @@
+	requeueAfter := deleteOlderThan + 1*time.Second
+	auditor := &discoveryAuditor{
+		deleteOlderThan: deleteOlderThan,
+		requeueAfter:    requeueAfter,
+	}
+
+	if createDiscoveryClient == nil {
+		createDiscoveryClient = func(ctx context.Context) (DiscoveryClient, error) {
+			return discovery.NewClient(ctx)
+		}
+	}
+
+	return qtransform.NewQController(
+		qtransform.Settings[*omni.ClusterIdentity, *omni.DiscoveryAuditResult]{
+			Name: "DiscoveryAuditController",
+			MapMetadataFunc: func(clusterIdentity *omni.ClusterIdentity) *omni.DiscoveryAuditResult {
+				return omni.NewDiscoveryAuditResult(resources.DefaultNamespace, clusterIdentity.Metadata().ID())
+			},
+			UnmapMetadataFunc: func(discoveryAuditResult *omni.DiscoveryAuditResult) *omni.ClusterIdentity {
+				return omni.NewClusterIdentity(resources.DefaultNamespace, discoveryAuditResult.Metadata().ID())
+			},
+			TransformFunc: func(ctx context.Context, _ controller.Reader, logger *zap.Logger, clusterIdentity *omni.ClusterIdentity, discoveryAuditResult *omni.DiscoveryAuditResult) error {
+				discoveryClient, err := createDiscoveryClient(ctx)
+				if err != nil {
+					return fmt.Errorf("failed to create discovery client: %w", err)
+				}
+
+				defer func() {
+					if closeErr := discoveryClient.Close(); closeErr != nil {
+						logger.Error("failed to close discovery client", zap.Error(closeErr))
+					}
+				}()
+
+				cluster := clusterIdentity.TypedSpec().Value.ClusterId
+
+				discoveryAuditResult.TypedSpec().Value.ClusterId = cluster
+
+				logger = logger.With(zap.String("cluster_id", cluster))
+
+				affiliatesInRemote, err := discoveryClient.ListAffiliates(ctx, cluster)
+				if err != nil {
+					return err
+				}
+
+				localAffiliates := clusterIdentity.TypedSpec().Value.NodeIds
+				pending, readyToDelete := auditor.calculateAffiliatesStatus(cluster, localAffiliates, affiliatesInRemote)
+
+				logger = logger.With(zap.Strings("pending", pending), zap.Strings("ready_to_delete", readyToDelete))
+
+				if len(readyToDelete) == 0 {
+					if len(pending) > 0 {
+						logger.Info("there are invalid affiliates that are not ready to be deleted yet, requeue the audit")
+
+						// there are invalid affiliates that are not ready to be deleted yet - requeue the audit
+						return controller.NewRequeueInterval(requeueAfter)
+					}
+
+					logger.Debug("no affiliates are ready to be deleted in this run, skip updating the result")
+
+					return xerrors.NewTaggedf[qtransform.SkipReconcileTag]("no affiliates to delete, skip updating the result")
+				}
+
+				logger.Info("found affiliates ready to be deleted")
+
+				deleted, requeue, err := auditor.deleteAffiliates(ctx, discoveryClient, logger, clusterIdentity, cluster, readyToDelete)
+				if err != nil {
+					return err
+				}
+
+				discoveryAuditResult.TypedSpec().Value.DeletedAffiliateIds = deleted
+
+				if requeue {
```

Comment:
Why don't we just remove affiliates that do not have corresponding `ClusterMachines` immediately?